test/e2e/operator_test.go

```diff
@@ -678,6 +614,124 @@ func TestExternalDNSSecretCredentialUpdate(t *testing.T) {
 	}
 }
 
+// TestExternalDNSAssumeRole tests the assumeRole functionality in which you can specify a Role ARN to use another
+// account's hosted zone for creating DNS records. Only AWS is supported.
+func TestExternalDNSAssumeRole(t *testing.T) {
+	// Only run this test if the DNS config contains the privateZoneIAMRole which indicates it's a "Shared VPC" cluster.
+	// Note: Only AWS supports privateZoneIAMRole.
+	dnsConfig := configv1.DNS{}
+	err := kubeClient.Get(context.TODO(), types.NamespacedName{Name: "cluster"}, &dnsConfig)
+	if err != nil {
+		t.Errorf("Failed to get dns 'cluster': %v\n", err)
+	}
+	if dnsConfig.Spec.Platform.AWS == nil || dnsConfig.Spec.Platform.AWS.PrivateZoneIAMRole == "" {
+		t.Skipf("Test skipped on non-shared-VPC cluster")
```

Comment:
Before I start a detailed review I would like to discuss the generic test/code organization.

The SharedVPC/AssumeRole support is different from other test cases. All the other test cases try to fit into the main pattern:
- create DNZ zone
- do platform agnostic tests
- delete DNS zone

In the case of SharedVPC I see some patterns that stand alone:
- DNZ zone is created by the CI
- DNZ zone is private
- Platform dependent - PrivateZoneRole is implemented only for AWS

I see that your code tries to balance in between "be like other test case" and "be specific". From one hand `providerTestHelper` interface is updated meaning that something is meant for all the providers while from the other hand the new test case is skipped for all the providers by AWS.

To make things more coherent I see 2 approaches:
1) Make an effort and generalize AssumeRole test case:
  -  add `helper.buildExternalDNSWithRole` method to `providerTestHelper` interface: AWS would get the role from the cluster's DNS config, other platforms would call the existing`buildExternalDNS()`.
  - keep the DNS record listing as part of  `providerTestHelper` interface but make the listing against different DNS zones: for AWS - from the other account, for the rest - from the same account.
  - `dig` probing can be kept as is
  
2) Move the new test case into a different file and update CI to run only the new test case in the shared-vpc CI job.

The first approach assumes that SharedVPC/AssumeRole features will come to the other providers too but makes a lot of unnecessary moves . The second approach is not generic but targets precisely the new feature.

Reply:
Good discussion. A lot of ways to proceed. I struggled a bit with this concept when writing this E2E. Couple of thoughts and questions:

- Sounds like the core question is whether it could ever be extended to other platforms. I see evidence (https://github.com/kubernetes-sigs/external-dns/issues/509) that folks are using service accounts to fulfill the assuming of a role in GCP.
    - It's possible that it could be extended, right? Or maybe there is a fundamental design difference that assume-role embedded in External DNS only makes sense in AWS.
- In addition to testing our new assume role functionality, is it reasonable to test non-assume-role functionality on a Shared VPC cluster?
   - At first, I thought this would a good test case, but the more I think about it, the more redundant it seems. Testing External DNS on a Shared VPC AWS cluster without using the Assume Role is not different than testing it in a non-Shared VPC vanilla AWS cluster.

> add helper.buildExternalDNSWithRole method to providerTestHelper interface: AWS would get the role from the cluster's DNS config, other platforms would call the existingbuildExternalDNS().

:+1: Except, I think you mean `AWS Shared VPC` instead of just `AWS`. Are you implying we should run the existing E2E tests under the `AWS Shared VPC` test job? It's possible, but messy. We are using a private hosted zone, so we'd need to change the logic to validate the test differently.

> keep the DNS record listing as part of providerTestHelper interface but make the listing against different DNS zones: for AWS - from the other account, for the rest - from the same account.

Are you suggesting to treat AWS Shared VPC as it's own platform (or provider)? i.e. add a new `awsAssumeRoleHelper` with a new file `aws_assume_role.go`? This could work, but we'd have to turn off the other non-assume-role E2E tests (like I mentioned above) and only run `TestExternalDNSAssumeRole` (due to private zone limitation).

I think my questions are:
1. Are you saying we should run the existing E2E tests under the `AWS Shared VPC` test job? And basically remove `TestExternalDNSAssumeRole` as a specific test?
2. Should `AWS Shared VPC` be treated as a it's own platform? I.e. in https://github.com/openshift/external-dns-operator/blob/829cd863bd5cbccdbcbb8f004d2c43fbd2ca229d/test/e2e/operator_test.go#L108-L120

Reply:
> Sounds like the core question is whether it could ever be extended to other platforms. I see evidence (Unable to use IAM Service Account on GKE kubernetes-sigs/external-dns#509) that folks are using service accounts to fulfill the assuming of a role in GCP.
>
>    It's possible that it could be extended, right? Or maybe there is a fundamental design difference that assume-role embedded in External DNS only makes sense in AWS.

Indeed, GCP seems to provide [SharedVPC](https://cloud.google.com/dns/docs/zones/zones-overview#cross-project_binding) feature too with a possibility to make [cross-projects DNS updates](https://github.com/kubernetes-sigs/external-dns/blob/master/docs/tutorials/gke.md#cross-project-access-scenario-using-google-service-account). However the way ExternalDNS is granted the permission seems to be different. The credentials are supposed to be from the other (not cluster's) project, they can be [packed as a secret](https://github.com/kubernetes-sigs/external-dns/blob/master/docs/tutorials/gke.md#static-credentials) and feed to ExternalDNS through [the credentials field](https://github.com/openshift/external-dns-operator/blob/main/api/v1beta1/externaldns_types.go#L288) because `CredentialsRequest` will provision credentials for the cluster's project. However I don't fully exclude the possibility that AssumeRole API will not be propagated (possibly in a different form) to other providers. I didn't see Azure and even GCP may be somehow adopted if not upstream  then in ExtDNS Operator.

> Except, I think you mean AWS Shared VPC instead of just AWS. Are you implying we should run the existing E2E tests under the AWS Shared VPC test job? It's possible, but messy. We are using a private hosted zone, so we'd need to change the logic to validate the test differently.

No, I meant AWS test helper.  Only the new test scenario can use the new `buildExternalDNSWithRole` method. The rest of the scenarios are still using the zone created the e2e test in `TestMain`. Even in the current implementation that is what's happening - all the old test cases are still using the public DNS zone from the cluster's account.

> Are you saying we should run the existing E2E tests under the AWS Shared VPC test job? And basically remove TestExternalDNSAssumeRole as a specific test?
> Should AWS Shared VPC be treated as a it's own platform?

Neither. The CI job you added is AWS platform (from the e2e point of view), it would run the same tests as any other platform. Only `TestExternalDNSAssumeRole` test case would be different - it will read the private DNZ zone ID and IAM role from the DNSConfig and return `ExternalDNS` with `AssumeRole` field filled in. The old AWS CI job would still run the old test plus the new `TestExternalDNSAssumeRole` but since there will be no private DNS zone ID or role in the dns config the `ExternalDNS` instance returned will not have `AssumeRole` field.
So this way the old AWS CI job and the one you added are almost the same except for 1 test case. Redundant but `TestExternalDNSAssumeRole` will fit into the test battery.


I think that I start to lean more towards the approach where your new CI job runs only 1 test - `TestExternalDNSAssumeRole` and this test is not even added to `operator_test.go` file but put into a dedicated "AWS specific" file. `TestExternalDNSAssumeRole` can be copied as is, it still can use all the utils and constants, it just doesn't need to be part of the framework build in `TestMain`.